code/dynamic_programming/src/digit_dp/digit_dp.py

```diff
@@ -0,0 +1,95 @@
+#################
+#  dynamic programming | digit dp | Python
+#  Part of Cosmos by OpenGenus Foundation
+#################
+
+# How many numbers x are there in the range a to b?
+# where the digit d occurs exactly k times in x?
+
+
+def digit_dp(n, d, k):
+	"""
+	Return total numbers less than equal to n having digit d in them k
+	times.
+	"""
+
+	# dp[pos][count][f] = Number of valid numbers <= b from this state
+	# pos = current position from left side (zero based)
+	# count = number of times we have placed the digit d so far
+	# f = the number we are building has already become smaller
+	# than b? [0 = no, 1 = yes]
+	dp = [[[-1] * 2 for i in range(k+1)] for i in range(len(str(n)))]
+
+	def digit_dp_helper(num, pos, count, f):
+		"""A recursive helper function for digit_dp function."""
+		if count > k:
+			return 0
+
+		if pos == len(num):
+			if count == k:
+				return 1
+			return 0
+
+		if dp[pos][count][f] != -1:
+			return dp[pos][count][f]
+
+		result = 0
+		limit = 0
+
+		if f == 0:
+			# Digits we placed so far matches with the prefix of b
+			# So if we place any digit > num[pos] in the current position, then
+			# the number will become greater than b
+			limit = int(num[pos]) - 0
+		else:
+			# The number has already become smaller than b.
+			# We can place any digit now.
+			limit = 9
+
+		# Try to place all the valid digits such that the number doesn't
+		# exceed b
+		for dgt in range(limit+1):
+			fNext = f
+			countNext = count
+
+			# The number is getting smaller at this position
+			if f == 0 and dgt < limit:
+				fNext = 1
+
+			if dgt == d:
+				countNext += 1
+
+			if countNext <= k:
+				result += digit_dp_helper(num, pos + 1, countNext, fNext)
+
+		dp[pos][count][f] = result
+		return dp[pos][count][f]
+
+	return digit_dp_helper(str(n), 0, 0, 0)
+
+
+def check_num(n, d, k):
+	"""Check if n has digit d in it k times."""
+	count = 0
+	for c in str(n):
+		if int(c) == d:
+			count += 1
+
+	if count == k:
+		return 1
+	else:
+		return 0
+
+
+def main():
+	"""Main function."""
+	a = 100
+	b = 100000000
+	d = 3
+	k = 5
+	ans = digit_dp(b, d, k) - digit_dp(a, d, k) + check_num(a, d, k)
```

Comment:
I think that to check a range `[a, b]`, we can simply compute `digit_dp(b, d, k) - digit_dp(a-1, d, k)`. This way we could get rid of the `check_num` function and make the code simpler.

Reply:
@Vitorvgc are you sure we need to do it this way? I wrote this code in C++, and the reason why I included the check() was because `a` and `b` are strings. And we cannot subtract an integer from a string in C++. Hence, had to add a check() to make sure I'm not missing out `a`.

But I don't know how it works in python. So please explain.

Reply:
I see, it is not necessary, I just think this way makes it simpler. Python also helps here. This change is easy to do here because the main function `digit_dp` already receives an integer, and inside of it there is a helper function `digit_dp_helper`, and only this helper function works with the number as a string. In the code we can see that in line 68 it is made a cast from int to string. So this way when calling the main function `digit_dp` we can work with integers, allowing to do `a-1`, and this function will call the recursive one that actually does the job, passing the number as string.

In your C++ code, you can achieve this by simply making the function receive an integer, and what is does is just cast it to string and call the DP function that works with the string. For example, you can change your `solve` function to receive an integer and then pass the string to the `digit_dp` function like this:

```c++
long solve(long num)
{
    memset(dp, -1, sizeof(dp));
    long result = digit_dp(to_string(num), 0, 0, 0);
    return result;
}
```

This way you could do:

```c++
int main() {
    long a = 100, b = 100000000;
    long ans = solve(b) - solve(a-1);
    cout << ans << endl;
}
```

And this will achieve the same result.

Reply:
I have a reason for using string. What if the number was a 2000 digit number?? It won't fit in any integer type. Hence used string.

Reply:
Yes, the drawback of using integer in this case is overflow problems. So if it needs to support arbitrary precision integers I think there is no way to avoid an extra function, either your check function or a "BigInteger" structure. Python makes things easier because it already supports arbitrary precision integers.

Reply:
Agree! @abhaygupta97 make the required changes.